File: 01-introduction-to-algorithms/c/01_binary_search.c

#include <stdio.h>

int binary_search(int[], int, int);

int binary_search(int myList[], int item, int len) {
  int low = 0;
  int high = len;

  while (low <= high) {
    int mid = (high + low) / 2;
    int guess = myList[mid];
    if (guess == item) {
      return mid;
    } else if (guess > item) {
      high = mid - 1;
    } else {
      low = mid + 1;
    }
  }

  return -1;
}

int main() {
  int myList[] = {1, 3, 5, 7, 9};
  int len = sizeof(myList) / sizeof(myList[0]);

  printf("%d\n", binary_search(myList, 3, len));  // 1
  printf("%d\n", binary_search(myList, -1, len)); //-1
  return 0;
}

File: 01-introduction-to-algorithms/go-lang/01_binary_search.go
package main

import "fmt"

func binary_search(myList []int, item int) int {
	low := 0
	high := len(myList) - 1

	for low <= high {
		mid := (low + high) / 2
		guess := myList[mid]

		if guess == item {
			return mid
		} else if guess > item {
			high = mid - 1
		} else {
			low = mid + 1
		}
	}

	return -1
}

func main() {
	myList := []int{1, 3, 5, 7, 9}
	fmt.Println(binary_search(myList, 3))  // 1
	fmt.Println(binary_search(myList, -1)) // -1
}
